### PR TITLE
Modified conformer with multi datasets

### PR DESCRIPTION
**Pull request by csukuangfj**
Use changes from
- https://github.com/k2-fsa/icefall/pull/288
- https://github.com/k2-fsa/icefall/pull/305

The training command is 
```
export CUDA_VISIBLE_DEVICES="0,1,2,3,4,5,6,7"

./pruned_transducer_stateless3/train.py \
  --world-size 8 \
  --num-epochs 20 \
  --start-epoch 0 \
  --full-libri 1 \
  --exp-dir pruned_transducer_stateless3/exp \
  --max-duration 300 \
  --use-fp16 1 \
  --lr-epochs 4 \
  --giga-prob 0.5
```

Note that the GigaSpeech dataset is not perturbed by speed and it is using on-the-fly feature extraction during training.
The XL subset (10k hours) together with the full LibriSpeech are used during training, with a batch being sampled with prob 0.5 from each subset.

---

Also, note that it does not support resuming training from a checkpoint saved in a sub-epoch.

---

EDITED: The tensorboard log is https://tensorboard.dev/experiment/lVZFGwjKS9iMYHzrsAE2cw/

### CODE REVIEW

**Comment by danpovey** [comment]
I believe you were having a problem with this being slow due to on-the-fly features from Gigaspeech?  Has that been resolved?

**Comment by danpovey** [comment]
I see that running this setup leads to commands like:
```
ffmpeg -threads 1 -ss 1803.05 -t 1.65 -i /ceph-fj/fangjun/open-source-2/icefall-pruned-multi-datasets/egs/librispeech/ASR/download/GigaSpeech/audio/podcast/P0032/POD0000003189.opus -ar 16000 -f f32le -threads 1 
```
running, but due to the -ss and -t options to ffmpeg, it does have to  decode the entire file.  I see the GPU-util is not great.  Sometimes, all but one machine is at 100% and one machine will be at 0% for 10 seconds or so (confusing)... sometimes they are all around 20%.

**Comment by pzelasko** [comment]
I don't believe that it has to decode the whole file when you specify offset and duration in ffmpeg; I was playing around with this in the past and got a great speedup by introducing these options.

**Comment by danpovey** [comment]
Yes, I tested the speed, it's fast.. sorry, my comment was a typo, I meant "doesn't" not "does".

File: egs/librispeech/ASR/pruned_transducer_stateless3/train.py [hunk 1]
```diff
@@ -0,0 +1,1123 @@
...
+        )
+    else:
+        cuts_musan = None
+
+    asr_datamodule = AsrDataModule(args)
+
+    train_dl = asr_datamodule.train_dataloaders(
+        train_cuts,
+        dynamic_bucketing=False,
+        on_the_fly_feats=False,
```
Comment (csukuangfj) [inline egs/librispeech/ASR/pruned_transducer_stateless3/train.py:982]: Note: The librispeech dataset is using pre-computed features so `torchaudio_load` should not be invoked

**Comment by csukuangfj** [comment]
Here are the results for far for this PR:

## greedy search
```
(py38) kuangfangjun:greedy_search$ grep -r -n --color "best for test-clean" log-* | sort -n -k2 | head
log-decode-epoch-17-avg-4-context-2-max-sym-per-frame-1-2022-04-17-22-46-50:15:greedy_search    2.47    best for test-clean
log-decode-epoch-17-avg-5-context-2-max-sym-per-frame-1-2022-04-17-22-47-45:15:greedy_search    2.47    best for test-clean
log-decode-epoch-17-avg-6-context-2-max-sym-per-frame-1-2022-04-17-22-48-42:15:greedy_search    2.48    best for test-clean
log-decode-epoch-17-avg-3-context-2-max-sym-per-frame-1-2022-04-17-22-45-56:15:greedy_search    2.49    best for test-clean
log-decode-epoch-16-avg-4-context-2-max-sym-per-frame-1-2022-04-17-21-43-34:15:greedy_search    2.5     best for test-clean
log-decode-epoch-16-avg-5-context-2-max-sym-per-frame-1-2022-04-17-21-44-33:15:greedy_search    2.5     best for test-clean
log-decode-epoch-17-avg-7-context-2-max-sym-per-frame-1-2022-04-17-22-49-41:15:greedy_search    2.51    best for test-clean
log-decode-epoch-16-avg-3-context-2-max-sym-per-frame-1-2022-04-17-21-42-41:15:greedy_search    2.52    best for test-clean
log-decode-epoch-16-avg-6-context-2-max-sym-per-frame-1-2022-04-17-21-45-34:15:greedy_search    2.53    best for test-clean
log-decode-epoch-17-avg-8-context-2-max-sym-per-frame-1-2022-04-17-22-50-40:15:greedy_search    2.53    best for test-clean
(py38) kuangfangjun:greedy_search$ grep -r -n --color "best for test-other" log-* | sort -n -k2 | head
log-decode-epoch-17-avg-7-context-2-max-sym-per-frame-1-2022-04-17-22-49-41:23:greedy_search    6.13    best for test-other
log-decode-epoch-17-avg-3-context-2-max-sym-per-frame-1-2022-04-17-22-45-56:23:greedy_search    6.15    best for test-other
log-decode-epoch-17-avg-5-context-2-max-sym-per-frame-1-2022-04-17-22-47-45:23:greedy_search    6.15    best for test-other
log-decode-epoch-17-avg-8-context-2-max-sym-per-frame-1-2022-04-17-22-50-40:23:greedy_search    6.15    best for test-other
log-decode-epoch-16-avg-4-context-2-max-sym-per-frame-1-2022-04-17-21-43-34:23:greedy_search    6.18    best for test-other
log-decode-epoch-17-avg-6-context-2-max-sym-per-frame-1-2022-04-17-22-48-42:23:greedy_search    6.18    best for test-other
log-decode-epoch-17-avg-4-context-2-max-sym-per-frame-1-2022-04-17-22-46-50:23:greedy_search    6.21    best for test-other
log-decode-epoch-17-avg-9-context-2-max-sym-per-frame-1-2022-04-17-22-51-40:23:greedy_search    6.21    best for test-other
log-decode-epoch-16-avg-5-context-2-max-sym-per-frame-1-2022-04-17-21-44-33:23:greedy_search    6.23    best for test-other
log-decode-epoch-16-avg-6-context-2-max-sym-per-frame-1-2022-04-17-21-45-34:23:greedy_search    6.23    best for test-other
```

## modified beam search
```
(py38) kuangfangjun:modified_beam_search$ grep -r -n --color "best for test-clean" log-* | sort -n -k2 | head
log-decode-epoch-17-avg-5-beam-4-2022-04-17-23-12-03:34:beam_size_4     2.4     best for test-clean
log-decode-epoch-17-avg-4-beam-4-2022-04-17-23-07-17:34:beam_size_4     2.41    best for test-clean
log-decode-epoch-16-avg-5-beam-4-2022-04-17-22-06-19:34:beam_size_4     2.43    best for test-clean
log-decode-epoch-17-avg-3-beam-4-2022-04-17-23-02-54:34:beam_size_4     2.43    best for test-clean
log-decode-epoch-17-avg-6-beam-4-2022-04-17-23-16-42:34:beam_size_4     2.43    best for test-clean
log-decode-epoch-16-avg-3-beam-4-2022-04-17-21-58-29:34:beam_size_4     2.44    best for test-clean
log-decode-epoch-16-avg-4-beam-4-2022-04-17-22-02-24:34:beam_size_4     2.44    best for test-clean
log-decode-epoch-17-avg-7-beam-4-2022-04-17-23-32-49:34:beam_size_4     2.44    best for test-clean
log-decode-epoch-16-avg-6-beam-4-2022-04-17-22-10-17:34:beam_size_4     2.45    best for test-clean
log-decode-epoch-17-avg-8-beam-4-2022-04-18-00-08-25:34:beam_size_4     2.45    best for test-clean
(py38) kuangfangjun:modified_beam_search$ grep -r -n --color "best for test-other" log-* | sort -n -k2 | head
log-decode-epoch-17-avg-7-beam-4-2022-04-17-23-32-49:61:beam_size_4     5.98    best for test-other
log-decode-epoch-17-avg-3-beam-4-2022-04-17-23-02-54:61:beam_size_4     6.0     best for test-other
log-decode-epoch-17-avg-5-beam-4-2022-04-17-23-12-03:61:beam_size_4     6.01    best for test-other
log-decode-epoch-17-avg-6-beam-4-2022-04-17-23-16-42:61:beam_size_4     6.01    best for test-other
log-decode-epoch-17-avg-8-beam-4-2022-04-18-00-08-25:61:beam_size_4     6.03    best for test-other
log-decode-epoch-16-avg-5-beam-4-2022-04-17-22-06-19:61:beam_size_4     6.04    best for test-other
log-decode-epoch-16-avg-6-beam-4-2022-04-17-22-10-17:61:beam_size_4     6.05    best for test-other
log-decode-epoch-17-avg-4-beam-4-2022-04-17-23-07-17:61:beam_size_4     6.05    best for test-other
log-decode-epoch-16-avg-4-beam-4-2022-04-17-22-02-24:61:beam_size_4     6.06    best for test-other
log-decode-epoch-15-avg-5-beam-4-2022-04-17-16-06-20:61:beam_size_4     6.07    best for test-other
```

Note: The WER for test-clean is 2.4 at epoch-17, --avg=5. It is expected to be lower as the training continues.

---

I will also try to decode the gigaspeech test datasets with the trained model.

**Comment by csukuangfj** [comment]
Here are the latest results for this PR.

## Giga prob 0.5

Also, Gigaspeech is speed-perturbed with factors 0.9 and 1.1

Training time per epoch:  about 2 hours

### greedy search
```
(py38) kuangfangjun:greedy_search$ grep -r -n --color "best for test-clean" log-* | sort -n -k2 | head
log-decode-epoch-29-avg-14-context-2-max-sym-per-frame-1-2022-04-19-12-10-16:15:greedy_search   2.31    best for test-clean
log-decode-epoch-30-avg-15-context-2-max-sym-per-frame-1-2022-04-19-11-55-20:15:greedy_search   2.32    best for test-clean
log-decode-epoch-33-avg-16-context-2-max-sym-per-frame-1-2022-04-19-22-56-07:15:greedy_search   2.32    best for test-clean
log-decode-epoch-29-avg-11-context-2-max-sym-per-frame-1-2022-04-19-12-07-07:15:greedy_search   2.33    best for test-clean
log-decode-epoch-29-avg-13-context-2-max-sym-per-frame-1-2022-04-19-12-09-12:15:greedy_search   2.33    best for test-clean
log-decode-epoch-32-avg-15-context-2-max-sym-per-frame-1-2022-04-19-15-04-25:15:greedy_search   2.33    best for test-clean
log-decode-epoch-27-avg-10-context-2-max-sym-per-frame-1-2022-04-19-13-46-24:15:greedy_search   2.34    best for test-clean
log-decode-epoch-28-avg-11-context-2-max-sym-per-frame-1-2022-04-19-12-20-48:15:greedy_search   2.34    best for test-clean
log-decode-epoch-28-avg-12-context-2-max-sym-per-frame-1-2022-04-19-12-21-50:15:greedy_search   2.34    best for test-clean
log-decode-epoch-28-avg-13-context-2-max-sym-per-frame-1-2022-04-19-12-22-56:15:greedy_search   2.34    best for test-clean
(py38) kuangfangjun:greedy_search$ grep -r -n --color "best for test-other" log-* | sort -n -k2 | head
log-decode-epoch-33-avg-17-context-2-max-sym-per-frame-1-2022-04-19-22-57-22:23:greedy_search   5.6     best for test-other
log-decode-epoch-30-avg-11-context-2-max-sym-per-frame-1-2022-04-19-11-45-16:23:greedy_search   5.62    best for test-other
log-decode-epoch-31-avg-15-context-2-max-sym-per-frame-1-2022-04-19-15-17-24:23:greedy_search   5.63    best for test-other
log-decode-epoch-32-avg-16-context-2-max-sym-per-frame-1-2022-04-19-15-05-34:23:greedy_search   5.63    best for test-other
log-decode-epoch-31-avg-12-context-2-max-sym-per-frame-1-2022-04-19-15-14-41:23:greedy_search   5.64    best for test-other
log-decode-epoch-32-avg-13-context-2-max-sym-per-frame-1-2022-04-19-15-02-19:23:greedy_search   5.64    best for test-other
log-decode-epoch-33-avg-14-context-2-max-sym-per-frame-1-2022-04-19-22-53-32:23:greedy_search   5.64    best for test-other
log-decode-epoch-33-avg-18-context-2-max-sym-per-frame-1-2022-04-19-22-58-37:23:greedy_search   5.64    best for test-other
log-decode-epoch-35-avg-20-context-2-max-sym-per-frame-1-2022-04-20-07-52-11:23:greedy_search   5.64    best for test-other
log-decode-epoch-30-avg-14-context-2-max-sym-per-frame-1-2022-04-19-11-48-11:23:greedy_search   5.65    best for test-other
```

### modified beam search
```
(py38) kuangfangjun:modified_beam_search$ grep -r -n --color "best for test-clean" log-* | sort -n -k2 | head
log-decode-epoch-29-avg-18-beam-4-2022-04-19-15-56-44:34:beam_size_4    2.29    best for test-clean
log-decode-epoch-28-avg-14-beam-4-2022-04-19-13-16-10:34:beam_size_4    2.3     best for test-clean
log-decode-epoch-28-avg-17-beam-4-2022-04-19-16-01-57:34:beam_size_4    2.3     best for test-clean
log-decode-epoch-29-avg-12-beam-4-2022-04-19-12-43-44:34:beam_size_4    2.3     best for test-clean
log-decode-epoch-29-avg-16-beam-4-2022-04-19-14-34-56:34:beam_size_4    2.3     best for test-clean
log-decode-epoch-29-avg-17-beam-4-2022-04-19-15-51-36:34:beam_size_4    2.3     best for test-clean
log-decode-epoch-30-avg-13-beam-4-2022-04-19-12-20-48:34:beam_size_4    2.3     best for test-clean
log-decode-epoch-32-avg-17-beam-4-2022-04-19-15-21-30:34:beam_size_4    2.3     best for test-clean
log-decode-epoch-28-avg-13-beam-4-2022-04-19-13-11-40:34:beam_size_4    2.31    best for test-clean
log-decode-epoch-28-avg-15-beam-4-2022-04-19-14-53-17:34:beam_size_4    2.31    best for test-clean
(py38) kuangfangjun:modified_beam_search$ grep -r -n --color "best for test-other" log-* | sort -n -k2 | head
log-decode-epoch-32-avg-18-beam-4-2022-04-19-15-26-18:61:beam_size_4    5.52    best for test-other
log-decode-epoch-30-avg-15-beam-4-2022-04-19-13-52-05:61:beam_size_4    5.53    best for test-other
log-decode-epoch-30-avg-14-beam-4-2022-04-19-12-29-45:61:beam_size_4    5.54    best for test-other
log-decode-epoch-31-avg-17-beam-4-2022-04-19-15-31-10:61:beam_size_4    5.54    best for test-other
log-decode-epoch-32-avg-17-beam-4-2022-04-19-15-21-30:61:beam_size_4    5.54    best for test-other
log-decode-epoch-28-avg-12-beam-4-2022-04-19-13-06-55:61:beam_size_4    5.55    best for test-other
log-decode-epoch-30-avg-16-beam-4-2022-04-19-14-02-53:61:beam_size_4    5.55    best for test-other
log-decode-epoch-28-avg-13-beam-4-2022-04-19-13-11-40:61:beam_size_4    5.56    best for test-other
log-decode-epoch-29-avg-13-beam-4-2022-04-19-12-48-18:61:beam_size_4    5.56    best for test-other
log-decode-epoch-29-avg-14-beam-4-2022-04-19-12-53-04:61:beam_size_4    5.56    best for test-other
```

## Giga prob 0.8 

Note: Gigapssech is not speed-perturbed.

Time per epoch: 5 hours 40 minutes

### greedy search
```
(py38) kuangfangjun:greedy_search$ grep -r -n --color "best for test-clean" log-* | sort -n -k2 | head
log-decode-epoch-19-avg-4-context-2-max-sym-per-frame-1-2022-04-25-12-01-30:15:greedy_search    2.32    best for test-clean
log-decode-epoch-20-avg-3-context-2-max-sym-per-frame-1-2022-04-25-17-13-40:15:greedy_search    2.32    best for test-clean
log-decode-epoch-20-avg-5-context-2-max-sym-per-frame-1-2022-04-25-17-15-53:15:greedy_search    2.32    best for test-clean
log-decode-epoch-19-avg-6-context-2-max-sym-per-frame-1-2022-04-25-12-04-07:15:greedy_search    2.33    best for test-clean
log-decode-epoch-20-avg-4-context-2-max-sym-per-frame-1-2022-04-25-17-14-38:15:greedy_search    2.33    best for test-clean
log-decode-epoch-20-avg-7-context-2-max-sym-per-frame-1-2022-04-25-17-18-30:15:greedy_search    2.33    best for test-clean
log-decode-epoch-20-avg-8-context-2-max-sym-per-frame-1-2022-04-25-17-19-38:15:greedy_search    2.33    best for test-clean
log-decode-epoch-18-avg-6-context-2-max-sym-per-frame-1-2022-04-25-10-40-00:15:greedy_search    2.34    best for test-clean
log-decode-epoch-19-avg-3-context-2-max-sym-per-frame-1-2022-04-25-12-00-18:15:greedy_search    2.34    best for test-clean
log-decode-epoch-20-avg-2-context-2-max-sym-per-frame-1-2022-04-25-17-12-28:15:greedy_search    2.34    best for test-clean
(py38) kuangfangjun:greedy_search$ grep -r -n --color "best for test-other" log-* | sort -n -k2 | head
log-decode-epoch-20-avg-5-context-2-max-sym-per-frame-1-2022-04-25-17-15-53:23:greedy_search    5.2     best for test-other
log-decode-epoch-20-avg-4-context-2-max-sym-per-frame-1-2022-04-25-17-14-38:23:greedy_search    5.21    best for test-other
log-decode-epoch-20-avg-7-context-2-max-sym-per-frame-1-2022-04-25-17-18-30:23:greedy_search    5.22    best for test-other
log-decode-epoch-20-avg-6-context-2-max-sym-per-frame-1-2022-04-25-17-17-19:23:greedy_search    5.23    best for test-other
log-decode-epoch-20-avg-3-context-2-max-sym-per-frame-1-2022-04-25-17-13-40:23:greedy_search    5.24    best for test-other
log-decode-epoch-19-avg-6-context-2-max-sym-per-frame-1-2022-04-25-12-04-07:23:greedy_search    5.25    best for test-other
log-decode-epoch-20-avg-2-context-2-max-sym-per-frame-1-2022-04-25-17-12-28:23:greedy_search    5.25    best for test-other
log-decode-epoch-20-avg-9-context-2-max-sym-per-frame-1-2022-04-25-17-21-02:23:greedy_search    5.25    best for test-other
log-decode-epoch-20-avg-8-context-2-max-sym-per-frame-1-2022-04-25-17-19-38:23:greedy_search    5.26    best for test-other
log-decode-epoch-19-avg-4-context-2-max-sym-per-frame-1-2022-04-25-12-01-30:23:greedy_search    5.27    best for test-other
```

### modified beam search
```
(py38) kuangfangjun:modified_beam_search$ grep -r -n --color "best for test-clean" log-* | sort -n -k2 | head
log-decode-epoch-20-avg-4-beam-8-2022-04-25-17-28-46:34:beam_size_8     2.28    best for test-clean
log-decode-epoch-20-avg-5-beam-8-2022-04-25-17-44-25:34:beam_size_8     2.28    best for test-clean
log-decode-epoch-19-avg-4-beam-4-2022-04-25-12-19-40:34:beam_size_4     2.3     best for test-clean
log-decode-epoch-18-avg-4-beam-4-2022-04-25-11-10-55:34:beam_size_4     2.31    best for test-clean
log-decode-epoch-18-avg-5-beam-4-2022-04-25-11-16-41:34:beam_size_4     2.31    best for test-clean
log-decode-epoch-18-avg-7-beam-4-2022-04-25-11-25-17:34:beam_size_4     2.31    best for test-clean
log-decode-epoch-19-avg-5-beam-4-2022-04-25-12-23-57:34:beam_size_4     2.31    best for test-clean
log-decode-epoch-19-avg-6-beam-4-2022-04-25-12-28-16:34:beam_size_4     2.31    best for test-clean
log-decode-epoch-18-avg-8-beam-4-2022-04-25-11-33-26:34:beam_size_4     2.32    best for test-clean
log-decode-epoch-19-avg-4-beam-8-2022-04-25-13-50-48:34:beam_size_8     2.32    best for test-clean
(py38) kuangfangjun:modified_beam_search$ grep -r -n --color "best for test-other" log-* | sort -n -k2 | head
log-decode-epoch-20-avg-5-beam-8-2022-04-25-17-44-25:61:beam_size_8     5.13    best for test-other
log-decode-epoch-20-avg-4-beam-8-2022-04-25-17-28-46:61:beam_size_8     5.15    best for test-other
log-decode-epoch-19-avg-6-beam-4-2022-04-25-12-28-16:61:beam_size_4     5.17    best for test-other
log-decode-epoch-19-avg-7-beam-4-2022-04-25-12-32-37:61:beam_size_4     5.2     best for test-other
log-decode-epoch-18-avg-3-beam-4-2022-04-25-10-55-21:61:beam_size_4     5.21    best for test-other
log-decode-epoch-19-avg-4-beam-8-2022-04-25-13-50-48:61:beam_size_8     5.21    best for test-other
log-decode-epoch-19-avg-5-beam-4-2022-04-25-12-23-57:61:beam_size_4     5.21    best for test-other
log-decode-epoch-19-avg-3-beam-4-2022-04-25-12-13-33:61:beam_size_4     5.22    best for test-other
log-decode-epoch-19-avg-4-beam-4-2022-04-25-12-19-40:61:beam_size_4     5.22    best for test-other
log-decode-epoch-19-avg-8-beam-4-2022-04-25-12-36-54:61:beam_size_4     5.22    best for test-other
```

You can see that using more data in training is helpful in reducing WER.

**Comment by danpovey** [comment]
Great!   So higher giga-prob seems to help more on test-other (like most things, I suppose).  These are good results.
We could probably  do even better if we used all of gigaspeech.

**Comment by csukuangfj** [comment]
Here are the final results for this PR
(See also https://github.com/csukuangfj/icefall/blob/modified-conformer-with-multi-datasets/egs/librispeech/ASR/RESULTS.md)

|                                     | test-clean | test-other | comment                                                                       |
|-------------------------------------|------------|------------|----------------------------------------|
| greedy search (max sym per frame 1) | 2.21       | 5.09       | --epoch 27 --avg 2  --max-duration 600 |
| greedy search (max sym per frame 1) | 2.25       | 5.02       | --epoch 27 --avg 12 --max-duration 600 |
| modified beam search                | 2.19       | 5.03       | --epoch 25 --avg 6  --max-duration 600 |
| modified beam search                | 2.23       | 4.94       | --epoch 27 --avg 10 --max-duration 600 |
| beam search                         | 2.16       | 4.95       | --epoch 25 --avg 7  --max-duration 600 |
| fast beam search                    | 2.21       | 4.96       | --epoch 27 --avg 10 --max-duration 600 |
| fast beam search                    | 2.19       | 4.97       | --epoch 27 --avg 12 --max-duration 600 |


The following table shows the
[Nbest oracle WER](http://kaldi-asr.org/doc/lattices.html#lattices_operations_oracle)
for fast beam search.
| epoch | avg | num_paths | nbest_scale | test-clean | test-other |
|-------|-----|-----------|-------------|------------|------------|
|  27   | 10  |   50      | 0.5         |  0.91      |  2.74      |
|  27   | 10  |   50      | 0.8         |  0.94      |  2.82      |
|  27   | 10  |   50      | 1.0         |  1.06      |  2.88      |
|  27   | 10  |   100     | 0.5         |  0.82      |  2.58      |
|  27   | 10  |   100     | 0.8         |  0.92      |  2.65      |
|  27   | 10  |   100     | 1.0         |  0.95      |  2.77      |
|  27   | 10  |   200     | 0.5         |  0.81      |  2.50      |
|  27   | 10  |   200     | 0.8         |  0.85      |  2.56      |
|  27   | 10  |   200     | 1.0         |  0.91      |  2.64      |
|  27   | 10  |   400     | 0.5         |  N/A       |  N/A       |
|  27   | 10  |   400     | 0.8         |  0.81      |  2.49      |
|  27   | 10  |   400     | 1.0         |  0.85      |  2.54      |

The Nbest oracle WER is computed using the following steps:

  - 1. Use `fast_beam_search` to produce a lattice.
  - 2. Extract `N` paths from the lattice using [k2.random_path](https://k2-fsa.github.io/k2/python_api/api.html#random-paths)
  - 3. [Unique](https://k2-fsa.github.io/k2/python_api/api.html#unique) paths so that each path
       has a distinct sequence of tokens
  - 4. Compute the edit distance of each path with the ground truth
  - 5. The path with the lowest edit distance is the final output and is used to
       compute the WER

The command to compute the Nbest oracle WER is:

```bash
for epoch in 27; do
  for avg in 10 ; do
    for num_paths in 50 100 200 400; do
      for nbest_scale in 0.5 0.8 1.0; do
        ./pruned_transducer_stateless3/decode.py \
            --epoch $epoch \
            --avg $avg \
            --exp-dir ./pruned_transducer_stateless3/exp \
            --max-duration 600 \
            --decoding-method fast_beam_search_nbest_oracle \
            --num-paths $num_paths \
            --max-states 32 \
            --beam 8 \
            --nbest-scale $nbest_scale
      done
    done
  done
done
```

Pretrained models, training logs, decoding logs, and decoding results
are available at
<https://huggingface.co/csukuangfj/icefall-asr-librispeech-pruned-transducer-stateless3-2022-04-29>


---

Note: The Nbest oracle WER shows that the WERs can be further reduced if some Nbest based LM rescoring models
are available.

---

I have tried to increase giga prob from 0.8 to 0.9. However, the training time per epoch is increased from 5 hours 40 minutes to 14 hours. It processes 12.5 k hours of training data for giga prob 0.8 and 30k hours of data.

Note: For giga prob 0.8, a new epoch is started as soon as the GigaSpeech dataset is exhausted, which is 10k hours of data.
For giga prob 0.9, the GigaSpeech dataset is repeated and will never be exhausted; a new epoch is started when the librispeech dataset is exhausted.

**Comment by csukuangfj** [comment]
I am merging it.

**Comment by danpovey** [comment]
cool!

**Pull request merged**